File: scripts/extract_novel.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
胄空形仙传 TXT章节提取脚本
从水印版TXT文件中提取完整的章节数据
"""

import re
import json
from pathlib import Path

# TXT源文件路径
TXT_FILE = r"e:\工作\清风\胄空形仙传（上）水印版.txt"
OUTPUT_FILE = r"d:\phpstudy_pro\WWW\aibot1\novel_chapters.json"

def read_file(path):
    """读取TXT文件"""
    with open(path, 'r', encoding='utf-8') as f:
        return f.read()

def parse_toc(content):
    """解析目录部分，提取章节列表"""
    chapters = []
    
    # 分割内容为行
    lines = content.split('\n')
    
    # 状态变量
    in_toc = False
    current_volume = "第一卷"
    chapter_id = 0
    
    # 用于匹配日期格式 2025.XX.XX
    date_pattern = re.compile(r'^(\d{4}\.\d{2}\.\d{2})$')
    # 用于匹配页码
    page_pattern = re.compile(r'^(\d+)$')
    
    # 临时存储
    temp_date = None
    temp_title = None
    temp_page = None
    
    # 遍历所有行
    i = 0
    while i < len(lines):
        line = lines[i].strip()
        
        # 检测卷次开始
        if line in ["第一卷", "第二卷", "第三卷", "番外篇"]:
            current_volume = line
            # 番外篇后续可能有编号，需要特殊处理
            if line == "番外篇" and i + 1 < len(lines):
                next_line = lines[i + 1].strip()
                if next_line.isdigit():
                    # 番外篇可能有子标题如"芽修传"
                    pass
            i += 1
            continue
        
        # 匹配日期
        date_match = date_pattern.match(line)
        if date_match:
            temp_date = date_match.group(1)
            # 继续获取标题
            if i + 1 < len(lines):
                next_line = lines[i + 1].strip()
                # 标题可能是多行的，需要处理
                if next_line and not page_pattern.match(next_line) and not date_pattern.match(next_line):
                    temp_title = next_line
                    i += 1
                    # 继续获取页码
                    if i + 1 < len(lines):
                        page_line = lines[i + 1].strip()
                        page_match = page_pattern.match(page_line)
                        if page_match:
                            temp_page = int(page_match.group(1))
                            # 找到完整记录，添加到章节列表
                            chapter_id += 1
                            chapters.append({
                                'id': chapter_id,
                                'date': temp_date,
                                'title': temp_title,
                                'volume': current_volume,
                                'page': temp_page
                            })
                            # 重置临时变量
                            temp_date = None
                            temp_title = None
                            temp_page = None
            i += 1
            continue
        
        i += 1
    
    return chapters

def extract_content_by_page(content, chapters):
    """根据页码提取章节内容"""
    # 分割内容为页
    pages = re.split(r'--- 第 (\d+) 页 ---', content)
    
    # pages[0] 是第一页之前的内容，pages[1]是第1页的页码，pages[2]是第1页的内容...
    page_contents = {}
    for i in range(1, len(pages), 2):
        page_num = int(pages[i])
        page_content = pages[i + 1] if i + 1 < len(pages) else ""
        page_contents[page_num] = page_content
    
    # 为每个章节填充内容
    for chapter in chapters:
        page_num = chapter.get('page')
        if page_num and page_num in page_contents:
            chapter['content'] = page_contents[page_num].strip()
        else:
            chapter['content'] = ""
    
    return chapters

def parse_toc_v2(content):
    """更智能地解析目录"""
    chapters = []
    
    lines = content.split('\n')
    
    current_volume = "第一卷"
    chapter_id = 0
    
    # 跳过前几页的封面和目录标题
    in_toc_section = False
    i = 0
    
    while i < len(lines):
        line = lines[i].strip()
        
        # 跳过空行
        if not line:
            i += 1
            continue
        
        # 检测卷次
        if line in ["第一卷", "第二卷", "第三卷"]:
            current_volume = line
            in_toc_section = True
            i += 1
            continue
        elif "番外篇" in line:
            current_volume = "番外篇"
            in_toc_section = True
            i += 1
            continue
        
        # 如果在目录区域，尝试匹配日期
        if in_toc_section:
            # 匹配日期格式 2025.XX.XX
            date_match = re.match(r'^(\d{4}\.\d{2}\.\d{2})$', line)
            if date_match:
                date = date_match.group(1)
                
                # 下一行是标题
                if i + 1 < len(lines):
                    title = lines[i + 1].strip()
                    # 跳过可能的标题为空的情况
                    if not title or re.match(r'^\d+$', title):
                        i += 1
                        continue
                    
                    # 再下一行是页码
                    if i + 2 < len(lines):
                        page_str = lines[i + 2].strip()
                        page_match = re.match(r'^(\d+)$', page_str)
                        if page_match:
                            chapter_id += 1
                            chapters.append({
                                'id': chapter_id,
                                'date': date,
                                'title': title,
                                'volume': current_volume,
                                'page': int(page_match.group(1))
                            })
                            i += 3
                            continue
                i += 1
                continue
        
        i += 1
    
    return chapters

def main():
    print("开始读取TXT文件...")
    content = read_file(TXT_FILE)
    print(f"文件总长度: {len(content)} 字符")
    
    print("\n解析目录...")
    chapters = parse_toc_v2(content)
    print(f"找到 {len(chapters)} 个章节")
    
    # 按卷次统计
    volumes = {}
    for ch in chapters:
        vol = ch['volume']
        volumes[vol] = volumes.get(vol, 0) + 1
    
    print("\n各卷章节数量:")
    for vol, count in volumes.items():
        print(f"  {vol}: {count} 章")
    
    # 显示前10个章节
    print("\n前10个章节:")
    for ch in chapters[:10]:
        print(f"  {ch['id']}. {ch['title']} ({ch['date']}) - 第{ch['page']}页")
    
    # 保存为JSON
    print(f"\n保存到 {OUTPUT_FILE}...")
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
        json.dump(chapters, f, ensure_ascii=False, indent=2)
    
    print("完成!")
    return chapters

if __name__ == "__main__":
    main()

File: scripts/extract_novel.js
#!/usr/bin/env node
/**
 * 胄空形仙传 TXT章节提取脚本 (完整版)
 * 从水印版TXT文件中提取所有章节数据
 */

const fs = require('fs');
const path = require('path');

// TXT源文件路径
const TXT_FILE = 'e:\\工作\\清风\\胄空形仙传（上）水印版.txt';
const OUTPUT_FILE = 'd:\\phpstudy_pro\\WWW\\aibot1\\novel_chapters.json';

function readFile(filePath) {
    return fs.readFileSync(filePath, 'utf8');
}

// 根据日期确定卷次
function getVolumeByDate(date) {
    if (date >= '2025.02.05' && date < '2025.03.03') return '第一卷';
    if (date >= '2025.03.03' && date < '2025.04.30') return '第二卷';
    if (date >= '2025.04.30' && date < '2025.07.01') return '第三卷';
    return '番外篇';
}

function parseToc(content) {
    const chapters = [];
    const lines = content.split('\n');
    
    let chapterId = 0;
    let pendingDate = null;
    
    // 匹配日期格式 2025.XX.XX
    const datePattern = /^(\d{4}\.\d{2}\.\d{2})$/;
    // 匹配页码（纯数字）
    const pagePattern = /^(\d+)$/;
    
    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();
        
        // 跳过空行
        if (!line) {
            pendingDate = null;
            continue;
        }
        
        // 跳过目录标题和页码标题行
        if (line === '目录' || line.includes('直播日期') || line.includes('当日内容简介') || line.includes('页码')) {
            continue;
        }
        
        // 检测卷次标记行 - 这些行只是标记，不作为章节
        if (line === '第一卷' || line === '第二卷' || line === '第三卷' || 
            (line.includes('番外篇') && !line.includes('芽修传') && !line.includes('（'))) {
            continue;
        }
        
        // 跳过特殊章节分类标题（如"芽修传"）
        if (line === '芽修传') {
            continue;
        }
        
        // 尝试匹配日期
        const dateMatch = line.match(datePattern);
        if (dateMatch) {
            pendingDate = dateMatch[1];
            
            // 继续查找标题
            if (i + 1 < lines.length) {
                const titleLine = lines[i + 1].trim();
                
                // 标题不能是纯数字或空
                if (titleLine && !pagePattern.test(titleLine) && titleLine.length > 1) {
                    // 继续查找页码
                    if (i + 2 < lines.length) {
                        const pageLine = lines[i + 2].trim();
                        const pageMatch = pageLine.match(pagePattern);
                        
                        if (pageMatch) {
                            const pageNum = parseInt(pageMatch[1], 10);
                            
                            // 正文从第6页开始
                            if (pageNum >= 6) {
                                chapterId++;
                                chapters.push({
                                    id: chapterId,
                                    date: pendingDate,
                                    title: titleLine,
                                    volume: getVolumeByDate(pendingDate),
                                    page: pageNum
                                });
                                
                                pendingDate = null;
                                i += 2; // 跳过标题和页码行
                                continue;
                            }
                        }
                    }
                }
            }
        }
        
        pendingDate = null;
    }
    
    return chapters;
}

function extractContentByPage(content, chapters) {
    // 按页分割内容
    const pageRegex = /--- 第 (\d+) 页 ---/g;
    const pageMatches = [...content.matchAll(pageRegex)];
    
    const pageContents = {};
    for (const match of pageMatches) {
        const pageNum = parseInt(match[1], 10);
        const startPos = match.index + match[0].length;
        
        // 找到下一页的开始位置
        let endPos = content.length;
        const nextMatch = pageMatches.find(m => m.index > startPos);
        if (nextMatch) {
            endPos = nextMatch.index;
        }
        
        pageContents[pageNum] = content.substring(startPos, endPos).trim();
    }
    
    // 为每个章节填充内容
    const pagePattern = /^(\d+)$/;
    const datePattern = /^(\d{4}\.\d{2}\.\d{2})$/;
    
    for (const chapter of chapters) {
        const pageNum = chapter.page;
        if (pageContents[pageNum]) {
            let rawContent = pageContents[pageNum];
            const contentLines = rawContent.split('\n');
            let contentStartIndex = 0;
            
            // 跳过前面的页码、日期、标题
            for (let j = 0; j < contentLines.length; j++) {
                const l = contentLines[j].trim();
                if (pagePattern.test(l) || datePattern.test(l)) {
                    contentStartIndex = j + 1;
                } else if (l && l !== chapter.title && l.length > 0 && !pagePattern.test(l)) {
                    break;
                }
            }
            
            chapter.content = contentLines.slice(contentStartIndex).join('\n').trim();
        } else {
            chapter.content = '';
        }
    }
    
    return chapters;
}

function main() {
    console.log('开始读取TXT文件...');
    const content = readFile(TXT_FILE);
    console.log(`文件总长度: ${content.length} 字符`);
    
    console.log('\n解析目录...');
    const chapters = parseToc(content);
    console.log(`找到 ${chapters.length} 个章节`);
    
    // 按卷次统计
    const volumes = {};
    for (const ch of chapters) {
        const vol = ch.volume;
        volumes[vol] = (volumes[vol] || 0) + 1;
    }
    
    console.log('\n各卷章节数量:');
    for (const [vol, count] of Object.entries(volumes)) {
        console.log(`  ${vol}: ${count} 章`);
    }
    
    // 显示所有章节
    console.log('\n所有章节列表:');
    let currentVol = '';
    for (const ch of chapters) {
        if (ch.volume !== currentVol) {
            currentVol = ch.volume;
            console.log(`\n=== ${currentVol} ===`);
        }
        console.log(`  ${ch.id}. ${ch.title} (${ch.date}) - 第${ch.page}页`);
    }
    
    // 提取内容
    console.log('\n\n提取章节内容...');
    const chaptersWithContent = extractContentByPage(content, chapters);
    
    // 保存为JSON
    console.log(`\n保存到 ${OUTPUT_FILE}...`);
    fs.writeFileSync(OUTPUT_FILE, JSON.stringify(chaptersWithContent, null, 2), 'utf8');
    
    console.log('完成!');
    
    // 统计内容长度
    let totalContentLength = 0;
    for (const ch of chaptersWithContent) {
        totalContentLength += (ch.content || '').length;
    }
    console.log(`\n总内容长度: ${totalContentLength} 字符`);
    console.log(`平均每章长度: ${Math.round(totalContentLength / chaptersWithContent.length)} 字符`);
    
    return chaptersWithContent;
}

main();
